### PR TITLE
feat: add pending transaction manual merging tool

### PR DESCRIPTION
**Pull request by AdamWHY2K**
Similar to the transaction matching tool, this PR allows users to select a pending transaction that has been posted but the auto detection hasn't picked up, then allows them to delete the pending transaction in favour of the posted one.

I've tested with lunchflow, but would require help testing plaid & simplefin.

### Screenshots

image

Screenshot_20260301_124157

### Copilot summary

This pull request introduces a new feature to allow users to manually merge a pending transaction with its posted version, addressing cases where automatic matching fails. It adds a dedicated UI for selecting and merging duplicates, updates the transaction model for maintainability, and includes comprehensive tests and localization.

**Manual Pending Duplicate Merge Feature**

* Added a new `PendingDuplicateMergesController` with `new` and `create` actions to handle the manual merge flow for pending transactions. This includes logic to select candidate posted transactions and perform the merge, updating metadata for auditability.
* Added a modal UI (`pending_duplicate_merges/new.html.erb`) to display the pending transaction, candidate posted transactions, and warnings, allowing users to select and confirm the merge.
* Added a call-to-action on the transaction show page for pending transactions, linking to the new manual merge modal.
* Added new routes for the pending duplicate merge resource under transactions.

**Model and Query Improvements**

* Refactored the `Transaction` model to centralize provider logic for pending transactions using a `PENDING_PROVIDERS` constant, simplifying the `pending`/`excluding_pending` scopes and `pending?` method. [[1]](diffhunk://#diff-162cad22ab0ca431166719b64d1f7bb2e883a0a63fd47a0219f367cec7f5f7dcR41-R53) [[2]](diffhunk://#diff-162cad22ab0ca431166719b64d1f7bb2e883a0a63fd47a0219f367cec7f5f7dcL81-R80)
* Added a `pending_duplicate_candidates` method to the `Transaction` model to efficiently find recent posted transactions from the same account and currency for manual matching.

**Localization and Testing**

* Added English translations for all new UI elements and warnings related to the manual merge feature. [[1]](diffhunk://#diff-551220dd7108ed817bdbd6623dab7bf324896e8d95c442fdac469bf9dfb284d8R1-R17) [[2]](diffhunk://#diff-1bdb25b7f95d9c0cd334ffd654d2cbadf7f20a4d916411dd0bd3b5c412f2ce25R43-R45) [[3]](diffhunk://#diff-1bdb25b7f95d9c0cd334ffd654d2cbadf7f20a4d916411dd0bd3b5c412f2ce25R94-R100)
* Added comprehensive controller tests for the new merge flow, including edge cases and model method coverage.

## Summary by CodeRabbit

* **New Features**
  * Merge pending transactions with posted transactions via a new “Merge with Posted Transaction” workflow.
  * In-UI modal and a transaction-show panel to review and select matching posted candidates with pagination.
  * Shows range and next/previous controls when browsing candidates; displays warnings and contextual details.

* **Localization**
  * Added English UI strings for the pending-duplicate merge flow.

* **Tests**
  * Added controller tests covering new flows, validations, and candidate filtering.

### CODE REVIEW

File: app/controllers/pending_duplicate_merges_controller.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+class PendingDuplicateMergesController < ApplicationController
+  before_action :set_transaction
+
+  def new
+    @limit = (params[:limit] || 20).to_i
+    @potential_duplicates = @transaction.pending_duplicate_candidates(limit: @limit)
+    # Check if there are more transactions available
+    @has_more = @transaction.pending_duplicate_candidates(limit: @limit + 1).count > @limit
```
[the remark this answers is not in the document]
Comment (AdamWHY2K) [reply app/controllers/pending_duplicate_merges_controller.rb]: I agree that this is a bad implementation, I don't agree with hard limiting how far back a user could merge a transaction though.

I'll fix the double query issue and switch to windowed pagination using an offset param which is sanitised to prevent negative values, but with no upper cap since the query is already scoped to a single account + currency with proper indexes, so it should be pretty performant even with large offsets.

This way users can still match old pending transactions without hitting an arbitrary limit.

File: app/controllers/pending_duplicate_merges_controller.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  def create
+    # Manually merge the pending transaction with the selected posted transaction
+    unless merge_params[:posted_entry_id].present?
+      redirect_back_or_to transactions_path, alert: "Please select a posted transaction to merge with"
+      return
+    end
+
+    posted_entry = Current.family.entries.find(merge_params[:posted_entry_id])
```
[the remark this answers is not in the document]
Comment (AdamWHY2K) [reply app/controllers/pending_duplicate_merges_controller.rb]: 7a7895a79abc522da9730874b8dc9eaab3538099

File: app/controllers/pending_duplicate_merges_controller.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        }
+      )
+    )
+
+    # Immediately merge
+    if @transaction.merge_with_duplicate!
+      redirect_back_or_to transactions_path, notice: "Pending transaction merged with posted transaction"
+    else
+      redirect_back_or_to transactions_path, alert: "Could not merge transactions"
+    end
```
[the remark this answers is not in the document]
Comment (AdamWHY2K) [reply app/controllers/pending_duplicate_merges_controller.rb]: 7a7895a79abc522da9730874b8dc9eaab3538099

File: app/controllers/pending_duplicate_merges_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdamWHY2K) [reply app/controllers/pending_duplicate_merges_controller.rb]: cbdb75d86e351dbe6fbcae612734fcf123912759 & 33b12b4499c031b9aec03ecda4ad803eabb9eedb

File: test/controllers/pending_duplicate_merges_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    post transaction_pending_duplicate_merges_path(pending_transaction), params: {
+      pending_duplicate_merges: {
+        posted_entry_id: posted_transaction.id
+      }
+    }
+
+    pending_transaction.reload
+    metadata = pending_transaction.entryable.extra["potential_posted_match"]
+
```
[the remark this answers is not in the document]
Comment (AdamWHY2K) [reply test/controllers/pending_duplicate_merges_controller_test.rb]: 6036506594b5ad41987f23b1d8a043f0adb19aea

**Comment by AdamWHY2K** [comment]
@jjmata pls could i get a review :pleading_face:

**Comment by jjmata** [comment]
Sure, I'm having a hard time keeping up!  Maybe @michelroegl-brunner can take a look/test this one out? 🙏

**Comment by michelroegl-brunner** [comment]
> Sure, I'm having a hard time keeping up!  Maybe @michelroegl-brunner can take a look/test this one out? 🙏 

Sure can, i report back my findings tomorrow ;)
